File: src/time_intersection_tool/time_intersection.py
"""
scheduler_pipeline.py

Core pipeline for multi-person free-time intersection:

1) normalize_calendars: parse user calendars into merged busy intervals
2) find_common_free: invert + intersect to get common free intervals
3) get_acceptable_times: placeholder to filter/rank free intervals (currently identity)

Assumes all events fall on the same date, and window provides that date context.
"""
from datetime import datetime, timedelta
from typing import List, Tuple, Dict
from functools import reduce

# Type alias for clarity
datetimeInterval = Tuple[datetime, datetime]

#------------------------
# Interval primitives
#------------------------

def merge(intervals: List[datetimeInterval]) -> List[datetimeInterval]:
    """Merge overlapping intervals into a sorted, non-overlapping list."""
    if not intervals:
        return []
    intervals.sort(key=lambda x: x[0])
    merged = [list(intervals[0])]
    for start, end in intervals[1:]:
        last = merged[-1]
        if start <= last[1]:
            last[1] = max(last[1], end)
        else:
            merged.append([start, end])
    return [(s, e) for s, e in merged]


def invert(busy: List[datetimeInterval], window: datetimeInterval) -> List[datetimeInterval]:
    """Subtract busy intervals from window to get free intervals."""
    free: List[datetimeInterval] = []
    cursor = window[0]
    for start, end in busy:
        if cursor < start:
            free.append((cursor, start))
        cursor = max(cursor, end)
    if cursor < window[1]:
        free.append((cursor, window[1]))
    return free


def intersect(a: List[datetimeInterval], b: List[datetimeInterval]) -> List[datetimeInterval]:
    """Intersect two lists of intervals."""
    i = j = 0
    out: List[datetimeInterval] = []
    while i < len(a) and j < len(b):
        start = max(a[i][0], b[j][0])
        end   = min(a[i][1], b[j][1])
        if start < end:
            out.append((start, end))
        if a[i][1] < b[j][1]:
            i += 1
        else:
            j += 1
    return out

#------------------------
# Pipeline functions
#------------------------

def normalize_calendars(
    profiles: List[Dict],
    window: datetimeInterval
) -> Dict[str, List[datetimeInterval]]:
    """
    Convert raw JSON profiles to a map of user -> merged busy intervals.
    Expects each profile {'name': str, 'calendar': [{'start': 'HH:MM','end': 'HH:MM', ...}, ...]}.
    All events are mapped onto window[0].date().
    """
    busy_map: Dict[str, List[datetimeInterval]] = {}
    for p in profiles:
        name = p['name']
        raw = []
        for ev in p.get('calendar', []):
            date = window[0].date()
            st = datetime.combine(date, datetime.strptime(ev['start'], '%H:%M').time())
            en = datetime.combine(date, datetime.strptime(ev['end'],   '%H:%M').time())
            raw.append((st, en))
        busy_map[name] = merge(raw)
    return busy_map


def find_common_free(
    busy_map: Dict[str, List[datetimeInterval]],
    window: datetimeInterval
) -> List[datetimeInterval]:
    """
    Compute common free intervals across all users in busy_map.
    """
    free_lists = []
    for intervals in busy_map.values():
        free_lists.append(invert(intervals, window))
    if not free_lists:
        return []
    return reduce(intersect, free_lists)


def get_acceptable_times(
    common_free: List[datetimeInterval]
) -> List[datetimeInterval]:
    """
    Placeholder: filter or rank 'common_free' intervals.
    Currently returns the full list unchanged.
    """
    return common_free

# Example usage
def example():
    # Define search window
    window_start = datetime(2025,8,5,9,0)
    window_end   = datetime(2025,8,5,17,0)
    window = (window_start, window_end)

    # Sample profiles
    profiles = [
        {'name':'Alice', 'calendar':[{'start':'12:00','end':'13:00'}, {'start':'14:00','end':'15:00'}]},
        {'name':'Bob',   'calendar':[{'start':'09:00','end':'13:00'}]}
    ]

    busy_map = normalize_calendars(profiles, window)
    common_free = find_common_free(busy_map, window)
    slots = get_acceptable_times(common_free)
    for s,e in slots:
        print(f"Free: {s.time()} → {e.time()}")

if __name__ == '__main__':
    example()

File: src/tools/time_intersection.ts
/**
 * @file time-intersection-tool.ts
 *
 * Core pipeline for multi-person free-time intersection, translated from Python.
 * This module is self-contained and uses native JavaScript Date objects for all calculations.
 *
 * 1) normalizeCalendars: Parses user calendars into merged busy intervals.
 * 2) findCommonFreeTime: Inverts and intersects busy times to get common free intervals within a specified time range.
 */

// --- Type Definitions ---

/** A tuple representing a start and end time for an interval. */
type DateTimeInterval = {
  start: Date,
  end: Date,
}

/** The expected structure for a user's profile and calendar data. */
export interface UserProfile {
  name: string
  calendar: {
    start: Date
    end: Date
    summary: string
  }[]
}

// --- Core Interval Primitives ---

/**
 * Merges overlapping and adjacent date intervals into a sorted, non-overlapping list.
 * @param intervals - A list of date intervals.
 * @returns A new list of merged intervals.
 */
function merge(intervals: DateTimeInterval[]): DateTimeInterval[] {
  if (intervals.length === 0) {
    return []
  }

  // Sort intervals by their start time
  intervals.sort((a, b) => a.start.getTime() - b.start.getTime())

  const merged: DateTimeInterval[] = [intervals[0]]

  for (let i = 1; i < intervals.length; i++) {
    const last = merged[merged.length - 1]
    const current = intervals[i]

    // If the current interval overlaps with the last one, merge them
    if (current.start <= last.end) {
      last.end = new Date(Math.max(last.end.getTime(), current.end.getTime()))
    } else {
      merged.push(current)
    }
  }
  return merged
}

/**
 * Subtracts a list of busy intervals from a given time window to find free intervals.
 * @param busy - A sorted, non-overlapping list of busy intervals.
 * @param window - The time window to find free time within.
 * @returns A list of free intervals.
 */
function invert(busy: DateTimeInterval[], window: DateTimeInterval): DateTimeInterval[] {
  const free: DateTimeInterval[] = []
  let cursor = window.start

  for (const interval of busy) {
    if (cursor < interval.start) {
      free.push({ start: cursor, end: interval.start })
    }
    cursor = new Date(Math.max(cursor.getTime(), interval.end.getTime()))
  }

  if (cursor < window.end) {
    free.push({ start: cursor, end: window.end })
  }

  return free
}

/**
 * Finds the intersection of two lists of sorted, non-overlapping intervals.
 * @param a - The first list of intervals.
 * @param b - The second list of intervals.
 * @returns A new list containing only the overlapping intervals.
 */
function intersect(a: DateTimeInterval[], b: DateTimeInterval[]): DateTimeInterval[] {
  const intersection: DateTimeInterval[] = []
  let i = 0
  let j = 0

  while (i < a.length && j < b.length) {
    const start = new Date(Math.max(a[i].start.getTime(), b[j].start.getTime()))
    const end = new Date(Math.min(a[i].end.getTime(), b[j].end.getTime()))

    if (start < end) {
      intersection.push({ start, end })
    }

    if (a[i].end < b[j].end) {
      i++
    } else {
      j++
    }
  }
  return intersection
}


// --- Pipeline Functions ---

/**
 * Converts raw user profiles into a map of user names to their merged busy intervals.
 * @param profiles - A list of user profiles.
 * @returns A dictionary mapping user names to their busy time intervals.
 */
function normalizeCalendars(profiles: UserProfile[]): Record<string, DateTimeInterval[]> {
  const busyMap: Record<string, DateTimeInterval[]> = {}

  for (const profile of profiles) {
    const rawIntervals: DateTimeInterval[] = profile.calendar.map((event) => ({
      start: event.start,
      end: event.end,
    }))
    busyMap[profile.name] = merge(rawIntervals)
  }

  return busyMap
}

// --- Main Exported Function ---

/**
 * The main function for the tool. It takes user profiles and calculates
 * common free time within the specified time range.
 * @param profiles - An array of user profiles with their calendar data.
 * @param startTime - The start of the time range to search for free slots.
 * @param endTime - The end of the time range to search for free slots.
 * @returns An array of common free time slots.
 */
export function findCommonFreeTime(profiles: UserProfile[], startTime: Date, endTime: Date): DateTimeInterval[] {
  if (!profiles || profiles.length === 0) {
    return []
  }

  const busyMap = normalizeCalendars(profiles)

  // Use the provided start and end times as the search window
  const window: DateTimeInterval = { start: startTime, end: endTime }

  // Invert each user's busy schedule to get their free time
  const freeLists = Object.values(busyMap).map((busyIntervals) => invert(busyIntervals, window))
  // Find the intersection of all users' free time
  // Start with the full window as the initial free time, then intersect with each user's free time
  const commonFree = freeLists.length > 0
    ? freeLists.reduce((acc, freeList) => intersect(acc, freeList), [window])
    : []

  return commonFree
}

// --- Conversion Functions ---

/**
 * Converts CalendarEvent array from the database schema (with ISO string dates)
 * to the format expected by UserProfile (with Date objects).
 * Filters out events where free=true (only keeps busy events).
 * @param events - Array of CalendarEvent objects from the database
 * @returns Array of calendar events with Date objects (only busy events)
 */
export function convertCalendarEventsToUserProfile(
  events: Array<{ start: string; end: string; summary: string; free?: boolean }>,
): Array<{ start: Date; end: Date; summary: string }> {
  return events
    .filter((event) => !event.free) // Only keep busy events (free !== true)
    .map((event) => ({
      start: new Date(event.start),
      end: new Date(event.end),
      summary: event.summary,
    }))
}

